cmd/natc/ippool/consensusippoolserialize.go

```diff
@@ -0,0 +1,181 @@
+func (ipp *ConsensusIPPool) Restore(rc io.ReadCloser) error {
+	//IPSet      *netipx.IPSet
+	//perPeerMap syncs.Map[tailcfg.NodeID, *consensusPerPeerState]
+	var snap fsmSnapshot
+	if err := json.NewDecoder(rc).Decode(&snap); err != nil {
+		return err
+	}
+	ipset, ppm, err := snap.getData()
+	if err != nil {
+		return err
+	}
+	ipp.IPSet = ipset
+	ipp.perPeerMap = ppm
+	return nil
+}
+
+type fsmSnapshot struct {
+	IPSet      persistableIPSet
+	PerPeerMap map[tailcfg.NodeID]persistablePPS
+}
+
+// Persist is part of the raft.FSMSnapshot interface
+// According to the docs Persist may be called concurrently with Apply
+func (f fsmSnapshot) Persist(sink raft.SnapshotSink) error {
+	b, err := json.Marshal(f)
+	if err != nil {
+		if marshalErr := sink.Cancel(); marshalErr != nil {
+			log.Printf("Error in sink.Cancel while handling error from json.Marshal: %v", err)
+		}
+		return err
+	}
+	if _, err := sink.Write(b); err != nil {
+		if cancelErr := sink.Cancel(); cancelErr != nil {
+			log.Printf("Error in sink.Cancel while handling error from sink.Write: %v", err)
+		}
+		return err
+	}
+	return sink.Close()
+}
+
+// Release is part of the raft.FSMSnapshot interface
+func (f fsmSnapshot) Release() {}
+
+// getPersistable returns an object that:
+// * contains all the data in ConsensusIPPool
+// * doesn't share any pointers with it
+// * can be marshalled to JSON
+// part of the raft snapshotting, getPersistable will be called during Snapshot
+// and the results used during persist (concurrently with Apply)
+func (ipp *ConsensusIPPool) getPersistable() fsmSnapshot {
+	ppm := map[tailcfg.NodeID]persistablePPS{}
+	for k, v := range ipp.perPeerMap.All() {
+		ppm[k] = v.getPersistable()
+	}
+	return fsmSnapshot{
+		IPSet:      getPersistableIPSet(ipp.IPSet),
+		PerPeerMap: ppm,
+	}
+}
+
+func (f fsmSnapshot) getData() (*netipx.IPSet, *syncs.Map[tailcfg.NodeID, *consensusPerPeerState], error) {
+	ppm := syncs.Map[tailcfg.NodeID, *consensusPerPeerState]{}
+	for k, v := range f.PerPeerMap {
+		ppm.Store(k, v.toPerPeerState())
+	}
+	ipset, err := f.IPSet.toIPSet()
+	if err != nil {
+		return nil, nil, err
+	}
+	return ipset, &ppm, nil
+}
+
+// getPersistable returns an object that:
+// * contains all the data in consensusPerPeerState
+// * doesn't share any pointers with it
+// * can be marshalled to JSON
+// part of the raft snapshotting, getPersistable will be called during Snapshot
+// and the results used during persist (concurrently with Apply)
+func (ps *consensusPerPeerState) getPersistable() persistablePPS {
+	dtaCopy := map[string]netip.Addr{}
+	for k, v := range ps.domainToAddr {
+		dtaCopy[k] = v
+	}
+	atd := map[netip.Prefix]whereWhen{}
+	for pfx, ww := range ps.addrToDomain.All() {
+		atd[pfx] = ww
+	}
+	return persistablePPS{
+		AddrToDomain: atd,
+		DomainToAddr: dtaCopy,
+	}
+}
+
+type persistablePPS struct {
+	DomainToAddr map[string]netip.Addr
+	AddrToDomain map[netip.Prefix]whereWhen
```

Comment:
It seems worthwhile to mention that bart.Table implements Cloner, and MarshalJSON and UnmarshalJSON. Handling this explicitly is fine, but it could also simply reuse the existing support.